Content.Server/Paper/PaperSystem.cs

```diff
@@ -1,20 +1,139 @@
+            UpdateUserInterface(uid, paperComp);
+        }
+
+        private void AfterUIOpen(EntityUid uid, PaperComponent paperComp, BeforeActivatableUIOpenEvent args)
+        {
+            paperComp.Mode = SharedPaperComponent.PaperAction.Read;
+            UpdateUserInterface(uid, paperComp);
+        }
+
+        private void OnExamined(EntityUid uid, PaperComponent paperComp, ExaminedEvent args)
+        {
+            if (!args.IsInDetailsRange)
+                return;
+
+            if (paperComp.Content != "")
+                args.Message.AddMarkup(
+                    Loc.GetString(
+                        "paper-component-examine-detail-has-words"
+                    )
+                );
+
+            if (paperComp.StampedBy.Count > 0)
+            {
+                args.Message.PushNewline();
+                string commaSeparated = string.Join(", ", paperComp.StampedBy);
+                args.Message.AddMarkup(
+                    Loc.GetString(
+                        "paper-component-examine-detail-stamped-by", ("stamps", commaSeparated))
+                );
+            }
         }
 
-        private void AfterUIOpen(EntityUid uid, PaperComponent component, BeforeActivatableUIOpenEvent args)
+        private void OnInteractUsing(EntityUid uid, PaperComponent paperComp, InteractUsingEvent args)
         {
-            component.Mode = SharedPaperComponent.PaperAction.Read;
-            component.UpdateUserInterface();
+            if (_tagSystem.HasTag(args.Used, "Write"))
+            {
+                if (!TryComp<ActorComponent>(args.User, out var actor))
+                return;
+
+                paperComp.Mode = PaperAction.Write;
+                UpdateUserInterface(uid, paperComp);
+                paperComp.UserInterface?.Open(actor.PlayerSession);
+                return;
+            }
+
+            if (TryComp<StampComponent>(args.Used, out var stampComp))
+            {
+                if (!paperComp.StampedBy.Contains(stampComp.StampedName))
+                {
+                    paperComp.StampedBy.Add(stampComp.StampedName);
+
+                    // If this is the first stamp, set appearance
+                    if (paperComp.StampedBy.Count == 1 && TryComp<AppearanceComponent>(uid, out var appearance))
+                        appearance.SetData(PaperVisuals.Stamped, true);
+                }
+
+                var stampPaperOtherMessage = Loc.GetString("paper-component-action-stamp-paper-other", ("user", args.User),("target", args.Target),("stamp", args.Used));
+                    _popupSystem.PopupEntity(stampPaperOtherMessage, args.User, Filter.Pvs(args.User).RemoveWhereAttachedEntity(puid => puid == args.User));
+                var stampPaperSelfMessage = Loc.GetString("paper-component-action-stamp-paper-self", ("target", args.Target),("stamp", args.Used));
+                    _popupSystem.PopupEntity(stampPaperSelfMessage, args.User, Filter.Entities(args.User));
```

Comment:
note to self: we really need helpers for these

Reply:
(Also please pass in EntityManager until we make filters a system)

Reply:
I added entitymanager where it was accepted in filter.pvs